File: studyUrl/src/main/java/com/surl/studyurl/domain/member/service/AuthTokenService.java
package com.surl.studyurl.domain.member.service;

import com.surl.studyurl.domain.member.entity.Member;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import com.surl.studyurl.global.app.AppConfig;
import java.security.SecureRandom;
import java.util.Base64;
import java.util.Date;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class AuthTokenService {
    public String genToken(Member member, long expireSeconds) {
        Claims claims = Jwts
                .claims()
                .add("id", member.getId())
                .add("userid", member.getUserid())

                .add("authorities", member.getAuthoritiesAsStringList())
                .build();

        Date issuedAt = new Date();
        Date expiration = new Date(issuedAt.getTime() + 1000 * expireSeconds);

        return Jwts.builder()
                .setClaims(claims)
                .setIssuedAt(issuedAt)
                .setExpiration(expiration)
                .signWith(SignatureAlgorithm.HS256, AppConfig.getJwtSecretKey())
                .compact();
    }

    public String genAccessToken(Member member) {
        return genToken(member, AppConfig.getAccessTokenExpirationSec());
    }

    public Map<String, Object> getDataFrom(String token) {
        Claims payload = Jwts.parser()
                .setSigningKey(AppConfig.getJwtSecretKey())
                .build()
                .parseClaimsJws(token)
                .getPayload();
        return Map.of(
                "id", payload.get("id", Integer.class),
                "userid", payload.get("userid", String.class),
                "authorities", payload.get("authorities", List.class)
        );
    }

    public boolean validateToken(String token) {
        try {
            Jwts.parser().setSigningKey(AppConfig.getJwtSecretKey()).build().parseClaimsJws(token);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public String genRefreshToken() {
        SecureRandom random = new SecureRandom();
        byte[] bytes = new byte[10];
        random.nextBytes(bytes);
        return Base64.getUrlEncoder().encodeToString(bytes);
    }
}

File: k_rstudy/src/main/kotlin/com/krstudy/kapi/domain/member/service/AuthTokenService.kt
package com.krstudy.kapi.domain.member.service

import com.krstudy.kapi.domain.member.entity.Member
import com.krstudy.kapi.global.app.AppConfig
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import org.springframework.stereotype.Service
import java.security.SecureRandom
import java.util.*

@Service
class AuthTokenService {

    fun genToken(member: Member, expireSeconds: Long): String {
        val claims = Jwts.claims()
            .add("id", member.id)
            .add("username", member.username)
            .add("authorities", member.getAuthoritiesAsStringList())
            .build()

        val issuedAt = Date()
        val expiration = Date(issuedAt.time + 1000 * expireSeconds)

        return Jwts.builder()
            .setClaims(claims)
            .setIssuedAt(issuedAt)
            .setExpiration(expiration)
            .signWith(SignatureAlgorithm.HS256, AppConfig.getJwtSecretKeyOrThrow())
            .compact()
    }

    fun genAccessToken(member: Member): String {
        return genToken(member, AppConfig.accessTokenExpirationSec) // 접근 방식 변경
    }

    fun getDataFrom(token: String): Map<String, Any> {
        val payload = Jwts.parser()
            .setSigningKey(AppConfig.getJwtSecretKeyOrThrow())
            .build()
            .parseClaimsJws(token)
            .body

        return mapOf(
            "id" to payload.get("id", Integer::class.java),
            "username" to payload.get("username", String::class.java),
            "authorities" to payload.get("authorities", List::class.java)
        )
    }

    fun validateToken(token: String): Boolean {
        return try {
            Jwts.parser().setSigningKey(AppConfig.getJwtSecretKeyOrThrow()).build().parseClaimsJws(token)
            true
        } catch (e: Exception) {
            false
        }
    }

    fun genRefreshToken(): String {
        val random = SecureRandom()
        val bytes = ByteArray(10)
        random.nextBytes(bytes)
        return Base64.getUrlEncoder().encodeToString(bytes)
    }
}